src/mscorlib/src/System/Lazy.cs

```diff
@@ -322,147 +398,116 @@ public T Value
-                    boxed = m_boxed as Boxed;
-                    if (boxed != null)
-                    {
-                        return boxed.m_value;
-                    }
-
-                    LazyInternalExceptionHolder exc = m_boxed as LazyInternalExceptionHolder;
-                    Debug.Assert(exc != null);
-                    exc.m_edi.Throw();
-                }
-
-                // Fall through to the slow path.
-                return LazyInitValue();
+                var implementation = m_implementation;
+                if (implementation == null)
+                    return m_value;
+                return implementation.Value;
             }
         }
 
-        /// <summary>
-        /// local helper method to initialize the value 
-        /// </summary>
-        /// <returns>The inititialized T value</returns>
-        private T LazyInitValue()
+        #region ILazyItem implementations
+
+        private sealed class LazyException : ILazyItem
         {
-            Boxed boxed = null;
-            LazyThreadSafetyMode mode = Mode;
-            if (mode == LazyThreadSafetyMode.None)
+            private readonly System.Runtime.ExceptionServices.ExceptionDispatchInfo m_exceptionInfo;
+            private readonly LazyThreadSafetyMode m_mode;
+
+            internal LazyException(Exception exception, LazyThreadSafetyMode mode)
             {
-                boxed = CreateValue();
-                m_boxed = boxed;
+                m_exceptionInfo = System.Runtime.ExceptionServices.ExceptionDispatchInfo.Capture(exception);
+                m_mode = mode;
             }
-            else if (mode == LazyThreadSafetyMode.PublicationOnly)
+
+            public bool IsValueCreated => false;
+            public bool IsValueFaulted => true;
+            public LazyThreadSafetyMode Mode => m_mode;
+
+            public T Value
             {
-                boxed = CreateValue();
-                if (boxed == null ||
-                    Interlocked.CompareExchange(ref m_boxed, boxed, null) != null)
-                {
-                    // If CreateValue returns null, it means another thread successfully invoked the value factory
-                    // and stored the result, so we should just take what was stored.  If CreateValue returns non-null
-                    // but another thread set the value we should just take what was stored.
-                    boxed = (Boxed)m_boxed;
-                }
-                else
+                get { m_exceptionInfo.Throw(); return default(T); }
+            }
+        }
+
+        abstract private class LazyInitializer : ILazyItem
+        {
+            protected Lazy<T> Owner { get; }
+            protected Func<T> Factory { get; private set; }
+
+            protected Func<T> TakeFactory()
+            {
+                var factory = Factory;
+                if (!ReferenceEquals(CreateInstance.Factory, factory))
```

Comment:
TakeFactory is required to handle re-enterency. In the case of recursive initialization it throws an exception. Needed to replicate existing functionality.